capsules/extra/src/date_time.rs

```diff
@@ -0,0 +1,354 @@
+        }
+    }
+
+    fn u32_as_dotw(&self, dotw_num: u32) -> Result<date_time::DayOfWeek, ErrorCode> {
+        match dotw_num {
+            0 => Ok(date_time::DayOfWeek::Sunday),
+            1 => Ok(date_time::DayOfWeek::Monday),
+            2 => Ok(date_time::DayOfWeek::Tuesday),
+            3 => Ok(date_time::DayOfWeek::Wednesday),
+            4 => Ok(date_time::DayOfWeek::Thursday),
+            5 => Ok(date_time::DayOfWeek::Friday),
+            6 => Ok(date_time::DayOfWeek::Saturday),
+            _ => Err(ErrorCode::INVAL),
+        }
+    }
+
+    /// Transforms two u32 numbers into a Date structure (year, month, dotm, dotw, hour, minute, seconds)
+    ///         first number (year, month, day_of_the_month):
+    ///                 -last 5 bits store the day_of_the_month
+    ///                 -previous 4 bits store the month
+    ///                 -previous 12 bits store the year
+    ///         second number (day_of_the_week, hour, minute, seconds):
+    ///                 -last 6 bits store the seconds
+    ///                 -previous 6 store the minute
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
+
+    fn date_from_u32_tuple(&self, date: u32, time: u32) -> Result<date_time::Date, ErrorCode> {
+        let month_num = date % (1 << 9) / (1 << 5);
+        let month_name = self.u32_as_month(month_num)?;
+
+        let dotw_num = time % (1 << 20) / (1 << 17);
+        let dotw_name = self.u32_as_dotw(dotw_num)?;
+
+        let date_result = date_time::Date {
+            year: (date % (1 << 21) / (1 << 9)) as u16,
+            month: month_name,
+            day: (date % (1 << 5)) as u8,
+
+            day_of_week: dotw_name,
+            hour: (time % (1 << 17) / (1 << 12)) as u8,
+            minute: (time % (1 << 12) / (1 << 6)) as u8,
+            seconds: (time % (1 << 6)) as u8,
+        };
+
+        if !(date_result.day <= 31) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.hour <= 24) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.minute <= 60) {
+            return Err(ErrorCode::INVAL);
+        }
+        if !(date_result.seconds <= 60) {
+            return Err(ErrorCode::INVAL);
+        }
+
+        Ok(date_result)
+    }
+
+    /// Transforms Date structure (year, month, dotm, dotw, hour, minute, seconds) into two u32 numbers
+    ///         first number (year, month, day_of_the_month):
+    ///                 -last 5 bits store the day_of_the_month
+    ///                 -previous 4 bits store the month
+    ///                 -previous 12 bits store the year
+    ///         second number (day_of_the_week, hour, minute, seconds):
+    ///                 -last 6 bits store the seconds
+    ///                 -previous 6 store the minute
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
```

Comment:
I think this should be moved to the top comment in the file since it's very important to how this driver works.